src/OrbitGl/CaptureWindowTest.cpp

```diff
@@ -161,113 +174,193 @@ class NavigationTestCaptureWindow : public CaptureWindow, public testing::Test {
+    }
   }
 };
 
-TEST_F(NavigationTestCaptureWindow, ZoomTimeWorksAsExpected) {
+TEST_F(NavigationTestCaptureWindow, ZoomTimeWorksInTheMiddleOfTimeGraph) {
+  PreRender();
   ExpectInitialState();
-
   const Vec2i kTimeGraphSize = viewport_.WorldToScreen(time_graph_->GetSize());
 
-  // Zoom time: Zoom in and out horizontally, assure we've zoomed in, and zoomed out
-  // again
-  // Test 1: Middle of the capture window
   int x = kTimeGraphSize[0] / 2;
-  int y = kTimeGraphSize[1] - 5;
+  int y = kTimeGraphSize[1] - kBottomSafetyMargin;
+
   MouseWheelMoved(x, y, 1, false);
-  ExpectIsHorizontallyZoomedIn(PosWithinCapture::kInBetween);
+  PreRender();
+  ExpectIsHorizontallyZoomedIn(PosWithinCapture::kMiddle);
+
   MouseWheelMoved(x, y, -1, false);
+  PreRender();
   ExpectInitialState();
 
   // Keyboard
   MouseMoved(x, y, false, false, false);
   KeyPressed('W', false, false, false);
-  ExpectIsHorizontallyZoomedIn(PosWithinCapture::kInBetween);
+  PreRender();
+  ExpectIsHorizontallyZoomedIn(PosWithinCapture::kMiddle);
+
   KeyPressed('S', false, false, false);
+  PreRender();
   ExpectInitialState();
+}
+
+TEST_F(NavigationTestCaptureWindow, ZoomTimeWorksAtRandomPositions) {
+  PreRender();
+  ExpectInitialState();
+  const Vec2i kTimeGraphSize = viewport_.WorldToScreen(time_graph_->GetSize());
+
+  std::random_device rd;
+  std::mt19937 mt(rd());
+  std::uniform_int_distribution<int> dist(0, kTimeGraphSize[0]);
+
+  constexpr int kNumTries = 100;
+  for (int i = 0; i < kNumTries; ++i) {
+    int x = dist(mt);
+    int y = kTimeGraphSize[1] - kBottomSafetyMargin;
+
+    // Zoom in twice, then zoom out twice. Check that the intermediate states match
+    MouseWheelMoved(x, y, 1, false);
+    PreRender();
+    ExpectIsHorizontallyZoomedIn(PosWithinCapture::kAnywhere);
+
+    float last_slider_pos = slider_->GetPosRatio();
+    double last_min_time = time_graph_->GetMinTimeUs();
+    double last_max_time = time_graph_->GetMaxTimeUs();
+    MouseWheelMoved(x, y, 1, false);
+    PreRender();
+    ExpectIsHorizontallyZoomedIn(PosWithinCapture::kAnywhere);
+
+    MouseWheelMoved(x, y, -1, false);
+    PreRender();
+    EXPECT_NEAR(slider_->GetPosRatio(), last_slider_pos, kSliderPosEpsilon);
+    EXPECT_NEAR(time_graph_->GetMinTimeUs(), last_min_time, kTimeEpsilonUs);
+    EXPECT_NEAR(time_graph_->GetMaxTimeUs(), last_max_time, kTimeEpsilonUs);
+
+    MouseWheelMoved(x, y, -1, false);
+    PreRender();
+    ExpectInitialState(true);
+
+    // Same test using the Keyboard
+    MouseMoved(x, y, false, false, false);
+    KeyPressed('W', false, false, false);
+    PreRender();
+    ExpectIsHorizontallyZoomedIn(PosWithinCapture::kAnywhere);
+
+    last_slider_pos = slider_->GetPosRatio();
+    last_min_time = time_graph_->GetMinTimeUs();
+    last_max_time = time_graph_->GetMaxTimeUs();
+    KeyPressed('W', false, false, false);
+    PreRender();
+    ExpectIsHorizontallyZoomedIn(PosWithinCapture::kAnywhere);
```

Comment:
I would be interested in testing conditions between pre-state and post-state of zooming in. We used to have an issue where zooming moved the screen for example.

I would like to provide the state before and check that min_timestamp is bigger or equal and max_timestamp is smaller or equal.

Reply:
Let's discuss this offline later and maybe add in a follow-up PR, I'm not sure if I understand this correctly.